### ISSUE TITLE
virtual line not rendering if it's the last line

### ISSUE THREAD

**Issue by dagle** [neovim/neovim#20179]
### Neovim version (nvim -v)

v0.8.0-dev-984-g568737d5b3

### Vim (not Nvim) behaves the same?

vim doesn't support the feature

### Operating system/version

archlinux

### Terminal name/version

foot version: 1.13.1

### $TERM environment variable

foot

### Installation

AUR

### How to reproduce the issue

```lua
local ns = vim.api.nvim_create_namespace("test-ns")
local opts = {
	virt_lines = {
		{{"text", "teststyle"}}
	},
}
vim.api.nvim_buf_set_extmark(0, ns, 7, 0, opts)
```

### Expected behavior

The virtual line should render because the line it's within the buffer.

### Actual behavior

The virtual line doesn't render at all and doesn't show any errors.

**Comment by zeertzjq** [comment]
The steps to reproduce are incomplete. What is the buffer content?

**Comment by zeertzjq** [comment]
Line 7 is already past the end of the buffer, so I think it is not clear what `virt_lines_above = false` means for that. You can add `virt_lines_above = true`, and it will be shown at the end of the buffer.

**Comment by luukvbaal** [comment]
I don't think this is a display issue? `nvim_buf_set_extmark()` is 0-indexed. The repro example places a `virt_line` at the 8th buffer line, which doesn't exist. The only problem here perhaps is that it doesn't show an error?

**Comment by zeertzjq** [comment]
It was decided somewhere in chat that this is a bug, but since this issue is four months old it may be hard to find the reference.

**Comment by luukvbaal** [comment]
I think some of the `ml_line_count` checks in `api/extmark.c:nvim_set_extmark()` should check for `ml_line_count - 1` instead? I.e. this patch (correctly?) shows an error for this repro:
```diff
+++ b/src/nvim/api/extmark.c
@@ -745,12 +745,12 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer line, Integer
   if (line < 0) {
     api_set_error(err, kErrorTypeValidation, "line value outside range");
     goto error;
-  } else if (line > buf->b_ml.ml_line_count) {
+  } else if (line >= buf->b_ml.ml_line_count) {
     if (strict) {
       api_set_error(err, kErrorTypeValidation, "line value outside range");
       goto error;
     } else {
-      line = buf->b_ml.ml_line_count;
+      line = buf->b_ml.ml_line_count - 1;
     }
   } else if (line < buf->b_ml.ml_line_count) {
     len = ephemeral ? MAXCOL : strlen(ml_get_buf(buf, (linenr_T)line + 1, false));
```

**Comment by zeertzjq** [comment]
Sorry, I mean it was decided somewhere in chat that the virtual line should be shown in this case.

**Comment by luukvbaal** [comment]
Oh okay, I wonder what the reason was then. Seems like an off-by-one error to me. (With the patch applied it would show the virtual line when `opts { strict == false }`).

**Comment by zeertzjq** [comment]
There may be some reason for extmarks one line beyond the end of the buffer to exist. If a `virt_lines_above` is added to the example in this issue the virtual line is then shown at the end of the buffer.
